File: built-in plugins dsp/mb-tremolo.cpp
/**
 * MB Tremolo
 * Category : effect
 * Type     : chorus
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Amplitude modulation
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_TREMOLO_H
#define MB_TREMOLO_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbTremolo : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-tremolo";
    static constexpr const char* PLUGIN_NAME    = "MB Tremolo";
    static constexpr const char* PLUGIN_TYPE    = "chorus";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float rate = 4f;  // range [0.5, 20]
    float depth = 0.5f;  // range [0, 1]
    float shape = 0.5f;  // range [0, 1]
    };

    MbTremolo() = default;
    ~MbTremolo() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.rate = std::clamp(params.rate, 0.5f, 20f);
        params.depth = std::clamp(params.depth, 0f, 1f);
        params.shape = std::clamp(params.shape, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Tremolo
        return input;
    }
};

#endif // MB_TREMOLO_H

File: built-in plugins dsp/mb-tremolo.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbTremoloPlugin: PluginDefinition = { id: 'mb-tremolo', slug: 'mb-tremolo', name: 'MB Tremolo', category: 'effect', type: 'chorus', version: '1.0.0', description: 'Amplitude modulation', author: 'Max Booster', parameters: [{ id: 'rate', name: 'Rate', type: 'float', defaultValue: 4, minValue: 0.5, maxValue: 20, automatable: true }, { id: 'depth', name: 'Depth', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }, { id: 'shape', name: 'Shape', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { rate: 4, depth: 0.5, shape: 0.5 } };

export default MbTremoloPlugin;
